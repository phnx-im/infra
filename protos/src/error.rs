use std::fmt;

use tonic::Status;

#[derive(Debug, thiserror::Error)]
#[error("missing field: {field}")]
pub struct MissingFieldError<E: fmt::Display> {
    field: E,
}

impl<E: fmt::Display> From<MissingFieldError<E>> for Status {
    fn from(e: MissingFieldError<E>) -> Self {
        Status::invalid_argument(e.to_string())
    }
}

pub trait MissingFieldExt<T, E: fmt::Display> {
    fn ok_or_missing_field(self, field: E) -> Result<T, MissingFieldError<E>>;
}

impl<T, E: fmt::Display> MissingFieldExt<T, E> for Option<T> {
    fn ok_or_missing_field(self, field: E) -> Result<T, MissingFieldError<E>> {
        self.ok_or(MissingFieldError { field })
    }
}

pub struct InvalidTls<F> {
    error: tls_codec::Error,
    field: F,
}

impl<F: fmt::Display> From<InvalidTls<F>> for Status {
    fn from(InvalidTls { error, field }: InvalidTls<F>) -> Status {
        Status::invalid_argument(format!("Invalid TLS field {field}: {error}"))
    }
}

pub struct TlsFailed<F> {
    error: tls_codec::Error,
    field: F,
}

impl<F: fmt::Display> From<TlsFailed<F>> for Status {
    fn from(TlsFailed { error, field }: TlsFailed<F>) -> Status {
        Status::internal(format!("TLS serialization of {field} failed: {error}",))
    }
}

pub trait InvalidTlsExt {
    type Value;

    fn invalid_tls<F>(self, field: F) -> Result<Self::Value, InvalidTls<F>>;

    fn tls_failed<F>(self, field: F) -> Result<Self::Value, TlsFailed<F>>;
}

impl<T> InvalidTlsExt for Result<T, tls_codec::Error> {
    type Value = T;

    fn invalid_tls<F>(self, field: F) -> Result<T, InvalidTls<F>> {
        self.map_err(|error| InvalidTls { error, field })
    }

    fn tls_failed<F>(self, field: F) -> Result<T, TlsFailed<F>> {
        self.map_err(|error| TlsFailed { error, field })
    }
}
