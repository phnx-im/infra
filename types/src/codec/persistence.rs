// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use serde::{Serialize, de::DeserializeOwned};
use sqlx::{Database, Decode, Encode, Type, encode::IsNull, error::BoxDynError};

use super::PhnxCodec;

pub struct BlobEncoded<T: Serialize>(pub T);

impl<DB: Database, T: Serialize> Type<DB> for BlobEncoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: Serialize> Encode<'q, DB> for BlobEncoded<T>
where
    Vec<u8>: Encode<'q, DB>,
{
    fn encode_by_ref(
        &self,
        buf: &mut <DB as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let bytes = PhnxCodec::to_vec(&self.0)?;
        Encode::<DB>::encode(&bytes, buf)
    }
}

#[derive(Debug)]
pub struct BlobDecoded<T>(pub T);

impl<T> BlobDecoded<T> {
    pub fn into_inner(self) -> T {
        self.0
    }
}

impl<DB: Database, T: DeserializeOwned> Type<DB> for BlobDecoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: DeserializeOwned> Decode<'q, DB> for BlobDecoded<T>
where
    for<'a> &'a [u8]: Decode<'a, DB>,
{
    fn decode(value: <DB as Database>::ValueRef<'q>) -> Result<Self, BoxDynError> {
        let bytes: &[u8] = Decode::<DB>::decode(value)?;
        PhnxCodec::from_slice(bytes)
            .map(BlobDecoded)
            .map_err(From::from)
    }
}
