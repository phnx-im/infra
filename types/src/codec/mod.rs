// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use cbor::Cbor;
use error::CodecError;
use mls_assist::memory_provider::Codec;
use serde::{Serialize, de::DeserializeOwned};

mod cbor;
mod error;
mod persistence;
#[cfg(test)]
mod tests;

pub use error::Error;
pub use persistence::{BlobDecoded, BlobEncoded};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(u8)]
#[non_exhaustive]
pub enum PhnxCodec {
    #[cfg(test)]
    OlderTestVersion = 0,
    #[default]
    V1 = 1,
}

impl TryFrom<u8> for PhnxCodec {
    type Error = Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            #[cfg(test)]
            0 => Ok(PhnxCodec::OlderTestVersion),
            1 => Ok(PhnxCodec::V1),
            _ => Err(Error::UnknownCodecVersion),
        }
    }
}

impl PhnxCodec {
    fn serialize_to_writer<T: Serialize>(
        &self,
        value: &T,
        writer: &mut impl std::io::Write,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
        // The first byte is always the codec version
        writer.write_all(&[*self as u8])?;
        match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::to_writer(value, writer)?,
            PhnxCodec::V1 => Cbor::to_writer(value, writer)?,
        }
        Ok(())
    }

    fn serialize<T: Sized + Serialize>(
        &self,
        value: &T,
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let mut buf = Vec::new();
        self.serialize_to_writer(value, &mut buf)?;
        Ok(buf)
    }

    fn deserialize<T: DeserializeOwned>(
        &self,
        bytes: &[u8],
    ) -> Result<T, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let res = match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::from_slice(bytes)?,
            PhnxCodec::V1 => Cbor::from_slice(bytes)?,
        };
        Ok(res)
    }

    pub fn to_vec<T>(value: &T) -> Result<Vec<u8>, Error>
    where
        T: Sized + Serialize,
    {
        let codec_version = PhnxCodec::default();
        let res = codec_version.serialize(value).map_err(|error| CodecError {
            codec_version,
            error,
        })?;
        Ok(res)
    }

    pub fn from_slice<T>(bytes: &[u8]) -> Result<T, Error>
    where
        T: DeserializeOwned,
    {
        let codec_version_byte = bytes.first().ok_or(Error::EmptyyInputSlice)?;
        let codec_version = PhnxCodec::try_from(*codec_version_byte)?;
        codec_version.deserialize(&bytes[1..]).map_err(|error| {
            CodecError {
                codec_version,
                error,
            }
            .into()
        })
    }
}

impl Codec for PhnxCodec {
    type Error = Error;

    fn to_vec<T>(value: &T) -> Result<Vec<u8>, Self::Error>
    where
        T: Sized + Serialize,
    {
        Self::to_vec(value)
    }

    fn from_slice<T>(bytes: &[u8]) -> Result<T, Self::Error>
    where
        T: DeserializeOwned,
    {
        Self::from_slice(bytes)
    }
}
