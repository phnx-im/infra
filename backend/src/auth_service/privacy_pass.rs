// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use async_trait::async_trait;
use phnxtypes::codec::{BlobDecoded, BlobEncoded};
use privacypass::{
    TruncatedTokenKeyId,
    batched_tokens_ristretto255::server::BatchedKeyStore,
    private_tokens::{Ristretto255, VoprfServer},
};
use sqlx::PgConnection;
use tokio::sync::Mutex;
use tracing::error;

pub(super) struct AuthServiceBatchedKeyStoreProvider<'a> {
    // Note: PgPool is not used here, because we need to use this provider in a transactional
    // context.
    connection: Mutex<&'a mut PgConnection>,
}

impl<'a> AuthServiceBatchedKeyStoreProvider<'a> {
    pub(super) fn new(connection: &'a mut PgConnection) -> Self {
        Self {
            connection: connection.into(),
        }
    }
}

#[async_trait]
impl BatchedKeyStore for AuthServiceBatchedKeyStoreProvider<'_> {
    /// Inserts a keypair with a given `truncated_token_key_id` into the key store.
    ///
    /// On conflict, an error is logged and the value is not inserted.
    async fn insert(
        &self,
        truncated_token_key_id: TruncatedTokenKeyId,
        server: VoprfServer<Ristretto255>,
    ) {
        let server = BlobEncoded(server);
        if let Err(error) = sqlx::query!(
            "INSERT INTO as_batched_keys (token_key_id, voprf_server)
            VALUES ($1, $2)",
            truncated_token_key_id as i16,
            server as _
        )
        .execute(&mut **self.connection.lock().await)
        .await
        {
            error!(%error, "Failed to insert key into batched key store");
        }
    }

    /// Returns a keypair with a given `truncated_token_key_id` from the key store.
    async fn get(
        &self,
        truncated_token_key_id: &TruncatedTokenKeyId,
    ) -> Option<VoprfServer<Ristretto255>> {
        let token_key_id: i16 = (*truncated_token_key_id).into();
        sqlx::query_scalar!(
            r#"SELECT voprf_server AS "voprf_server: BlobDecoded<VoprfServer<Ristretto255>>"
            FROM as_batched_keys
            WHERE token_key_id = $1"#,
            token_key_id
        )
        .fetch_optional(&mut **self.connection.lock().await)
        .await
        .inspect_err(|error| error!(%error, "Failed to fetch key from batched key store"))
        .ok()?
        .map(|BlobDecoded(voprf_server)| voprf_server)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::LazyLock;

    use phnxtypes::codec::PhnxCodec;
    use rand::{SeedableRng, rngs::StdRng};
    use sqlx::PgPool;

    use super::*;

    #[sqlx::test]
    async fn insert_get(pool: PgPool) -> anyhow::Result<()> {
        let mut connection = pool.acquire().await?;
        let provider = AuthServiceBatchedKeyStoreProvider::new(&mut connection);

        let mut rng = rand::thread_rng();

        let value = VoprfServer::new(&mut rng).unwrap();
        provider.insert(1, value.clone()).await;

        let loaded = provider.get(&1).await.unwrap();
        assert_eq!(loaded, value);

        let value = VoprfServer::new(&mut rng).unwrap();
        provider.insert(2, value.clone()).await;

        let loaded = provider.get(&2).await.unwrap();
        assert_eq!(loaded, value);

        Ok(())
    }

    #[sqlx::test]
    async fn no_insert_on_conflict(pool: PgPool) -> anyhow::Result<()> {
        let mut connection = pool.acquire().await?;
        let provider = AuthServiceBatchedKeyStoreProvider::new(&mut connection);

        let mut rng = rand::thread_rng();

        let value_a = VoprfServer::new(&mut rng).unwrap();
        provider.insert(1, value_a.clone()).await;

        let loaded = provider.get(&1).await.unwrap();
        assert_eq!(loaded, value_a);

        let value_b = VoprfServer::new(&mut rng).unwrap();
        provider.insert(1, value_b.clone()).await;

        let loaded = provider.get(&1).await.unwrap();
        assert_eq!(loaded, value_a);

        Ok(())
    }

    static SERVER: LazyLock<VoprfServer<Ristretto255>> = LazyLock::new(|| {
        VoprfServer::new(&mut StdRng::seed_from_u64(0x0DDB1A5E5BAD5EEDu64)).unwrap()
    });

    #[test]
    fn test_server_serde_codec() {
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&*SERVER).unwrap());
    }

    #[test]
    fn test_server_serde_json() {
        insta::assert_json_snapshot!(&*SERVER);
    }
}
