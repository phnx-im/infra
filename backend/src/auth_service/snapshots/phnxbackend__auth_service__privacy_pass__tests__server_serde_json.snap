---
source: backend/src/auth_service/privacy_pass.rs
expression: "&*SERVER"
---
{
  "sk": [
    195,
    107,
    56,
    47,
    5,
    197,
    27,
    126,
    203,
    132,
    212,
    200,
    213,
    127,
    233,
    146,
    152,
    14,
    64,
    118,
    185,
    228,
    218,
    135,
    208,
    2,
    113,
    44,
    18,
    135,
    173,
    1
  ],
  "pk": [
    136,
    249,
    199,
    209,
    224,
    210,
    127,
    224,
    69,
    6,
    72,
    17,
    91,
    51,
    235,
    15,
    74,
    161,
    191,
    41,
    218,
    18,
    201,
    212,
    219,
    68,
    97,
    133,
    247,
    190,
    225,
    53
  ]
}
